Integration of FuelTypeName enums & Removal of functions about FuelTypeNames

dareumnam opened the pull request:
Pull request overview
---------------------
 - Integrated a few separate enum classes regarding fuel type names
   - `enum class eResource`: ResourceType (with everything)
   - `enum class eFuel`: FuelType (only fuels)
   - `enum class ePollutant`: Pollutant (only pollutants)
   - Added std::arrays to map between these enums: 
     - `std::array<eFuel, static_cast<int>eResource::Num> eResource2eFuel`
     - `std::array<eResource, static_cast<int>eFuel::Num> eFuel2eResource`
     - `std::array<ePollutant, static_cast<int>eResource::Num> eResource2ePollutant`
     - `std::array<eResource, static_cast<int>ePollutant::Num> ePollutant2eResource`
   - Added Names and NamesUC Arrays
 - Removed `ValidateFuelTypeWithAssignResourceTypeNum()`
 - Removed `AssignResourceTypeNum()`
 - Removed `GetResourceTypeChar()`
 - Revised `GetStandardMeterResourceType()`
 - Removed `DataGlobalConstants.cc`
 - Removed `DataGlobalConstants.unit.cc` 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Myoldmopar committed:
Remove unused optional arg in ValidateFuelType

Myoldmopar committed:
Remove unused state argument from ValidateFuelType

Myoldmopar committed:
Convert code to use an enum and getNumerationValue instead of ValidateFuelType

Myoldmopar committed:
Convert more code to use the same enum for fuel type and remove the other validate function

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into SCV_UtilityRoutines

dareumnam committed:
pull SCV_UtilityRoutines branch and merge conflicts

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Rename enums

dareumnam committed:
create ResourceTypeNames arrays, remove AssignResourceTypeNum()

dareumnam committed:
pull develop, merge conflict, and address comments

dareumnam committed:
combine UtilityRoutines::FuelType1 with DataGlobalConstants::ResourceType

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Remove GetResourceTypeChar()

dareumnam committed:
Remove DataGlobalConstants.cc and DataGlobalConstants.unit.cc

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
remove WaterThermalTanks::Fuel

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
break pollutants out of ResourceType enum

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
move pollutants to PollutionModules

dareumnam committed:
comment unused variables out to remove warnings

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -102,24 +101,7 @@ namespace DataGlobalConstants {
-        NuclearLow,
-        WaterEnvironmentalFactors,
         CarbonEquivalent,
-        Source,
```
Can someone explain to me what physical thing `Source` is?

mjwitte replied:
Source energy, i.e. the amount of energy used to produce a unit of energy at the meter. e.g. How much gas energy went into the power plant to produce a kWh of electricity at the meter (including losses, etc.). This should probably be renamed to `SourceEnergy`.

amirroth replied:
Aha. That makes sense. Strange that SourceEnergy is so aggregated whereas everything else is so disaggregated.

mjwitte replied:
It's the original measure from long ago for making apples-to-apples comparisons of electricity vs gas (or other end-use fuel types). $ are another proxy, and now CarbonEquivalent is a biggie, which is aggregated back to the source as well.

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -185,9 +226,6 @@ namespace DataGlobalConstants {
 
-    ResourceType AssignResourceTypeNum(std::string const &ResourceTypeChar);
-    std::string GetResourceTypeChar(ResourceType ResourceTypeNum);
-
```
Good bye!

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -210,14 +248,14 @@ struct DataGlobalConstantsData : BaseGlobalStruct
     static int constexpr iEndUseSize = 14;
 
-    std::vector<DataGlobalConstants::ResourceType> AllResourceTypes = {DataGlobalConstants::ResourceType::None,
+    std::vector<DataGlobalConstants::ResourceType> AllResourceTypes = {DataGlobalConstants::ResourceType::Invalid,
```
Should get rid of this vector in general. Just use the enumeration itself.

amirroth commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -190,7 +190,7 @@ namespace DataZoneEquipment {
 
         // Default Constructor
         EquipMeterData()
```
Initialize in declaration, remove explicit default constructor. If you touch an object, you should do this.

amirroth commented on src/EnergyPlus/EconomicLifeCycleCost.hh:
```diff
@@ -364,7 +364,7 @@ namespace EconomicLifeCycleCost {
 
         // Default Constructor
-        UsePriceEscalationType() : resource(DataGlobalConstants::ResourceType::None), escalationStartYear(0), escalationStartMonth(0)
+        UsePriceEscalationType() : resource(DataGlobalConstants::ResourceType::Invalid), escalationStartYear(0), escalationStartMonth(0)
```
Same comment.

amirroth commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -2379,20 +2379,17 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
-            UtilityRoutines::ValidateFuelTypeWithAssignResourceTypeNum(
-                cAlphaArgs(39), state.dataHVACVarRefFlow->VRF(VRFNum).FuelType, state.dataHVACVarRefFlow->VRF(VRFNum).FuelTypeNum, FuelTypeError);
-            if (FuelTypeError) {
+            state.dataHVACVarRefFlow->VRF(VRFNum).FuelTypeNum = static_cast<DataGlobalConstants::ResourceType>(
```
These variables should just be `fuelType`. A `Type` (i.e., an enumeration) is a number, so `TypeNum` is redundant.

amirroth commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -2544,7 +2541,6 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
         state.dataHVACVarRefFlow->VRF(VRFNum).Name = cAlphaArgs(1);
         state.dataHVACVarRefFlow->VRF(VRFNum).VRFSystemTypeNum = VRF_HeatPump;
         state.dataHVACVarRefFlow->VRF(VRFNum).VRFAlgorithmTypeNum = AlgorithmType::FluidTCtrl;
-        state.dataHVACVarRefFlow->VRF(VRFNum).FuelType = "Electricity";
```
Bye bye now.

amirroth commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -5160,40 +5157,40 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
                             state.dataHVACVarRefFlow->VRF(NumCond).Name);
         SetupOutputVariable(state,
-                            "VRF Heat Pump Cooling " + state.dataHVACVarRefFlow->VRF(NumCond).FuelType + " Rate",
+                            format("VRF Heat Pump Cooling {} Rate", sFuelType),
```
👍

amirroth commented on src/EnergyPlus/HVACVariableRefrigerantFlow.hh:
```diff
@@ -240,12 +242,11 @@ namespace HVACVariableRefrigerantFlow {
-        Real64 TUHeatingLoad;                               // total TU heating load for each VRF system
-        bool SwitchedMode;                                  // used to derate capacity/power when system changes operating mode
+        DataGlobalConstants::ResourceType FuelTypeNum = DataGlobalConstants::ResourceType::Invalid; // Fuel type number
+        Real64 SUMultiplier;                                                                        // exponential timer for mode changes
```
Initialize the rest of these in the declaration also. Yowzers, there's a lot of them. So maybe not.

amirroth commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -7541,7 +7542,8 @@ void GetMeteredVariables(EnergyPlusData &state,
 
-            ResourceTypes.at(NumVariables) = AssignResourceTypeNum(UtilityRoutines::MakeUPPERCase(op->EnergyMeters(MeterPtr).ResourceType));
+            ResourceTypes.at(NumVariables) = static_cast<DataGlobalConstants::ResourceType>(
+                getEnumerationValue(DataGlobalConstants::ResourceTypeNamesUC, UtilityRoutines::MakeUPPERCase(op->EnergyMeters(MeterPtr).ResourceType)));
```
I will deal with you later, young man! 📌

mjwitte replied:
If these are the only changes in OutputProcessor, then it seems like this PR breaks the possibility of reporting pollutants as meters. I'll make a test file to confirm. Oh, I see, [`GetStandardMeterResourceType`](https://github.com/NREL/EnergyPlus/blob/7cc2fa5a6f00d9f5db1cc54d853f11ea74851582/src/EnergyPlus/OutputProcessor.cc#L1515-L1541) has not been touched so it does not use `DataGlobalConstants::ResourceTypes` to validate yet. So, when it is changed over someday, it will have to check 2 enums, `ResourceTypes` and `PollutionsModule::Pollutant` (should that be `PollutantTypes`?) to cover all of the valid meter resources. But that's where this may fall apart, when the resource type key in `SetUpOutputVariable` is changed to an enum, that enum will need to include the union of `ResourceType` and `PollutantTypes`.

amirroth replied:
Is this the *only* reason that these disparate enums are thrown together? To simplify OutputProcessor? That was a sub-optimal design choice. It's better to complicate OutputProcessor and simplify the remaining 98% of the code than the reverse.

mjwitte replied:
> Is this the _only_ reason that these disparate enums are thrown together? To simplify OutputProcessor? That was a sub-optimal design choice. It's better to complicate OutputProcessor and simplify the remaining 98% of the code than the reverse.

I'm not sure. We'd have to go back to see the history of when the `ResourceType` enum was first introduced. It may have been part of a big enum sweep without recognizing the interaction with other functions like the ones being eliminated here.

amirroth replied:
I will point out that in OutputProcessor specifically, ResourceType is a `string`, not an `enum`. That's not great in general but at least in this case it makes it simpler to deal with multiple enums as a single set in a compartmentalized way. The next question is whether the `ResourceType enum` is used anywhere other than as `FuelType`. From this PR, it looks like the answer is no.

mjwitte replied:
I don't claim to understand fully how OutputProcessor is managing meters and submeters, but the string name is important because that ultimately becomes part of the meter name like "Electricity:Facility". Ultimately I believe each metered value ends up being the sum of a list of pointers to the respective variables that feed into it, so at that point it doesn't care about ResourceType anymore, that's just a part of the particular meter's name.

Ultimately, when calling `SetupOutputVariable` I think we would prefer that `ResourceTypeKey` was an enum rather than a string. That's where the full combined enum would be needed, and then it would be promptly used to grab the string version to create the concatenated meter name. After that, I'm not sure if OutputProcessor uses the resource type explicitly or not.

amirroth replied:
In that case, how would you suggest we handle the myriad cases where we only need the `FuelType` subset and the one case where we only need the `Pollutant` subset? Use separate enumerations for those and use `std::array`s to map one to the other?

mjwitte replied:
Maybe. `Pollutants` are a different beast, because all of the pollutants are active at the same time and the enum could be used to cycle through arrays by pollutant. So I think you really need a dedicated enum for the pollution module, but then, yes you want a way to look up the equivalent `ResourceType` enum value to pass into `SetupOutputVariable`.

`FuelType` on the other hand is used differently. Any given component, like a boiler, has a single main fuel type, so here, looking it up and storing it as a `ResourceType` is good, because it's only used to set up meters and name output variables. So, for this, perhaps there could be a companion bool array, say `IsValidFuelType` which is true or false for each ResourceType and then after `getenumeration` returns the `ResourceType` check 
```
if (IsValueFuelType[static_cast<int>Boiler.ResourceType])
```
or something like that?

amirroth replied:
OutputReportTabular defines `numResourceTypes (14)` and `numSourceTypes (12)` as the first 14 and 12 items in `ResourceType` (the difference between the two is that `SourceTypes` does not include district heating and cooling). The former especially is used as in the index in a large number of arrays. How would you handle this use?

amirroth replied:
> `FuelType` on the other hand is used differently. Any given component, like a boiler, has a single main fuel type, so here, looking it up and storing it as a `ResourceType` is good, because it's only used to set up meters and name output variables. So, for this, perhaps there could be a companion bool array, say `IsValidFuelType` which is true or false for each ResourceType and then after `getenumeration` returns the `ResourceType` check.
> 
> ```
> if (IsValueFuelType[static_cast<int>Boiler.ResourceType])
> ```
> 
> or something like that?

That's a possibility, and I grant you that there are different kinds of beasts. But `FuelType` is also used as array indices (or at least it should be instead of what is used now which is just the numbers 1, 2, 3, etc.) and there is something to be said for having a consistent way of doing things. I have an idea, I will talk to @dareumnam about it and we will see how it works.

amirroth commented on src/EnergyPlus/UtilityRoutines.cc:
```diff
@@ -453,131 +453,6 @@ namespace UtilityRoutines {
         }
     }
 
-    bool ValidateFuelType([[maybe_unused]] EnergyPlusData &state,
```
Happy happy day!

Myoldmopar replied:
No kidding!

amirroth commented on src/EnergyPlus/WaterThermalTanks.cc:
```diff
@@ -10872,25 +10878,25 @@ void WaterThermalTankData::SizeTankForDemandSide(EnergyPlusData &state)
         // assume can propagate rules for gas to other fuels.
         bool FuelTypeIsLikeGas = false;
         switch (this->FuelType) {
```
Don't use control flow to set boolean values. Just do 

```cpp
FuelTypeIsLikeGas = 
   (this->FuelType == DataGlobalConstants::ResourceType::NaturalGas || 
    this->FuelType == ...);
```

Another option is to do a `std::array<bool, ResourceType::Num> IsFuelTypeLikeGas = { false, true, ... }'` and then do lookups. That might better.

mjwitte commented on src/EnergyPlus/WaterThermalTanks.hh:
```diff
@@ -204,50 +204,6 @@ namespace WaterThermalTanks {
         Num
     };
 
-    enum class Fuel
```
@dareumnam Have we lost the distinction between resource types and fuel types (which are a subset of resource types)? We've had this same issue in other places of the code where one enum class is a subset of a larger group, but we've kept the seemingly duplicate enums.

amirroth replied:
I believe the idea is to restore that distinction, but to do it only once globally. And we need to discuss how to do this because there are various `Fuel` subsets. It was easier to break `Pollutant` out, because that list is used in a much more limited way and is not ambiguous. (Except for `Source` which I am not sure is a pollutant, it seems to me like `Source` should be a variable of type `Fuel` rather than its own type. And then there would also be a `TransmissionLoss` variable. This stuff is wacky).

mjwitte replied:
The term "ResourceType" and the full master list was for anything that can be metered. (mtd/mtr ouputs) It was perhaps a misnomer when things like pollutants were added to the list of things you can meter in EnergyPlus. It combined a bunch of overlapping sublists, fuel types, energy flows, electricity meter flows, pollutants, water stuff, etc. And it gave the outputprocessor a single "list" to work with for valid resource types when creating meters. I need to look more closely at the big picture of this PR.

amirroth replied:
I know what you are saying and I agree that we need to reconsider this approach. However, I don't think we should hold up this PR to do so. In general, for things like this, it's good to have smaller "cleanup" PRs, especially if they result in as many textual changes as this one does (you will see this with an upcoming Material cleanup PR from @yujiex). This PR is conceptually small (except for the breakout of Pollutant, but that appears to be non-controversial yes?), but physically quite large. Reconsidering `ResourceType` will be a bigger conceptual PR and we don't want to review the cross-product of that change with this one.

mjwitte commented on src/EnergyPlus/Boilers.cc:
```diff
@@ -210,18 +210,14 @@ void GetBoilerInput(EnergyPlusData &state)
-        UtilityRoutines::ValidateFuelTypeWithAssignResourceTypeNum(
-            state.dataIPShortCut->cAlphaArgs(2), thisBoiler.BoilerFuelTypeForOutputVariable, thisBoiler.FuelType, FuelTypeError);
-        if (FuelTypeError) {
+        thisBoiler.FuelType = static_cast<DataGlobalConstants::ResourceType>(
+            getEnumerationValue(DataGlobalConstants::ResourceTypeNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(2))));
```
Based on the boiler (haven't looked at other components) the code here would allow anything in the resource type list to be a valid fuel to power the boiler. The only place the limits the choices here is the key choice list in the IDD/schema for this input field. We've relied on that in other places to weed out nonsense inputs but usually the key choice list and the enum are the same list. Here we're relying on the IDD/schema to restrict to a subset. I suppose that's ok. But if you're just reading the source code here, you don't know that the full list of `ResourceType` is not valid for this object.

amirroth replied:
🤔 Maybe we do want to consider this in this PR after all.

dareumnam committed:
clang format

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
separate enum classes

dareumnam commented:
To-do list for this branch:

 - [X] To have three enum ’s: 
1. `enum class eResource`: ResourceType (with everything)
2. `enum class eFuel`: FuelType (only fuels)
3. `enum class ePollutant`: Pollutant (only pollutants)


 - [x] To map between them with four maps: 
1. `std::array<eFuel, static_cast<int>eResource::Num> eResource2eFuel = {eFuel::Electricity, eFuel::NaturalGas, … eFuel::Invalid, eFuel::Invalid};`
2. And then the reverse:
`std::array<eResource, static_cast<int>eFuel::Num> eFuel2eResource = {eResource::Electricity, eResource::NaturalGas, …}`
3. and 4. same for Pollutant.


 - [x] As far as Names and NamesUC:
1. `eResourceNames` is defined as usual. 
2. `eFuelNames` is defined as
`constexpr std::array<std::string_view, static_cast<int>eFuel::Num> eFuelNames = {eResourceNames[static_cast<int>eFuel2eResource[eFuel::Electricity]], eResourceNames[static_cast<int>eFuel2eResource[eFuel::NaturalGas]], …};`
so that the names are kept in sync.

nrel-bot-2 commented:
@dareumnam @myoldmopar it has been 28 days since this pull request was last updated.

dareumnam committed:
Pull develop and merge conflicts

dareumnam committed:
fix errors

dareumnam committed:
clang format

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
fix error

dareumnam committed:
add exceptions in enum custom check:

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
add std::arrays to map between Resource and Fuel/Pollutant, rename arrays

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam committed:
add exceptions in custom check & clang format

dareumnam committed:
add exceptions in custom check enum scope

dareumnam committed:
Remove GetStandardMeterResourceType() and Add General to eResource

dareumnam committed:
Add initialization

dareumnam committed:
Revert "Add initialization"

This reverts commit a647d3f2db7224d23e8ae009e5473a4b2206c506.

dareumnam committed:
Revert "Remove GetStandardMeterResourceType() and Add General to eResource"

This reverts commit 283766807f435d21d455e522de67cf3521fe656d.

dareumnam committed:
revised GetStandardMeterResourceType()

dareumnam commented:
This is ready for review. @Myoldmopar @amirroth @mjwitte 

This PR
- Integrated a few separate enum classes regarding fuel type names
   - `enum class eResource`: ResourceType (with everything)
   - `enum class eFuel`: FuelType (only fuels)
   - `enum class ePollutant`: Pollutant (only pollutants)
   - Added std::arrays to map between these enums: 
     - `std::array<eFuel, static_cast<int>eResource::Num> eResource2eFuel`
     - `std::array<eResource, static_cast<int>eFuel::Num> eFuel2eResource`
     - `std::array<ePollutant, static_cast<int>eResource::Num> eResource2ePollutant`
     - `std::array<eResource, static_cast<int>ePollutant::Num> ePollutant2eResource`
   - Added Names and NamesUC Arrays
 - Removed `ValidateFuelTypeWithAssignResourceTypeNum()`
 - Removed `AssignResourceTypeNum()`
 - Removed `GetResourceTypeChar()`
 - Revised `GetStandardMeterResourceType()`
 - Removed `DataGlobalConstants.cc`
 - Removed `DataGlobalConstants.unit.cc`

amirroth commented on scripts/dev/check_for_enum_scope_usage.py:
```diff
@@ -197,10 +197,13 @@ def run(self):
-                unique_files_in_usages.add(u.file_path.name)
-            if len(unique_files_in_usages) == 1:
-                apparent_enums_in_only_one_source_file.append(f"{e.describe()} in {next(iter(unique_files_in_usages))}")
+            # exceptions listed by <FILE>:<ENUM NAME>
```
If people object to this new convention of beginning enumerations with `e` (I prefer it to ending all enumerations with `Type`) I can get rid of it using a `sed` script after this PR is merged (or give the `sed` script to @dareumnam).

jasondegraw replied:
I object, for the same reasons I objected to introducing Hungarian notation. IDEs do a really good job now of telling what is what, so in my opinion encoding type information in the name of the variable isn't needed.

amirroth replied:
How do you feel about 40% of the enums in EnergyPlus ending in the word `Type`?

jasondegraw replied:
I'm not a huge fan, for me it falls into the same category as all the uses of `Data` as a prefix/suffix.

mjwitte replied:
I don't have a strong opinion on naming conventions, but why change the pattern now and then add exceptions to the script that's enforcing the existing rule?

amirroth commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -196,15 +197,13 @@ namespace BoilerSteam {
-            UtilityRoutines::ValidateFuelTypeWithAssignResourceTypeNum(
-                state.dataIPShortCut->cAlphaArgs(2), thisBoiler.BoilerFuelTypeForOutputVariable, thisBoiler.FuelType, FuelTypeError);
-            if (FuelTypeError) {
+            thisBoiler.FuelType = static_cast<Constant::eResource>(
```
Not sure if we want to do this in this PR or not, but would be good to convert all of these FuelType variables from `eResource` to `eFuel`.

mjwitte replied:
> Not sure if we want to do this in this PR or not, but would be good to convert all of these FuelType variables from `eResource` to `eFuel`.

@dareumnam @amirroth  I must be missing something. It appears that the subset enums like `eFuel` are not being used anywhere. They are declared and mapped to each other in `DataConstants` but I don't see them used elsewhere. 

In this case here, `ValidateFuelTypeWithAssignResourceTypeNum` rejected any input that was not a fuel type, but now, using `getEnumerationValue(Constant::eResourceNamesUC . . .` will accept any resource type. The key choices in the IDD will reject any non-fuels, but do we want to rely on that being the only gatekeeper?

dareumnam replied:
@mjwitte Thanks for your insight. As you said, `eFuel` has not been used yet in this PR, but I also think we should convert all FuelType variables from `eResource` to `eFuel`, not just rely on InputProcessor to reject non-fuels. After this PR, I'll open a new PR for that.

dareumnam replied:
Or I can do that in this PR either.

mjwitte replied:
> @mjwitte Thanks for your insight. As you said, `eFuel` has not been used yet in this PR, but I also think we should convert all FuelType variables from `eResource` to `eFuel`, not just rely on InputProcessor to reject non-fuels. After this PR, I'll open a new PR for that.

> Or I can do that in this PR either.

@dareumnam Given the amount of changes here, I would finish this PR and start a new one. Be sure to explain this in the top comment on the PR.

amirroth commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -196,15 +197,13 @@ namespace BoilerSteam {
-                state.dataIPShortCut->cAlphaArgs(2), thisBoiler.BoilerFuelTypeForOutputVariable, thisBoiler.FuelType, FuelTypeError);
-            if (FuelTypeError) {
+            thisBoiler.FuelType = static_cast<Constant::eResource>(
+                getEnumerationValue(Constant::eResourceNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(2))));
```
Do not need `MakeUPPERCase` for things that come from `InputProcessor`, it already uppercases everything.

amirroth commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -196,15 +197,13 @@ namespace BoilerSteam {
-            if (FuelTypeError) {
+            thisBoiler.FuelType = static_cast<Constant::eResource>(
+                getEnumerationValue(Constant::eResourceNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(2))));
+            if (thisBoiler.FuelType == Constant::eResource::Invalid) {
```
Also, this check is not needed if the field is required because the epJSON parser catches this type of thing already (so you will never get here).

amirroth commented on src/EnergyPlus/Boilers.cc:
```diff
@@ -207,16 +207,14 @@ void GetBoilerInput(EnergyPlusData &state)
-        if (FuelTypeError) {
+        thisBoiler.FuelType = static_cast<Constant::eResource>(
+            getEnumerationValue(Constant::eResourceNamesUC, UtilityRoutines::MakeUPPERCase(state.dataIPShortCut->cAlphaArgs(2))));
+        if (thisBoiler.FuelType == Constant::eResource::Invalid) {
```
Similarly not needed.

amirroth commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.hh:
```diff
@@ -55,6 +55,7 @@
 #include <EnergyPlus/DataGlobalConstants.hh>
 #include <EnergyPlus/DataLoopNode.hh>
 #include <EnergyPlus/EnergyPlus.hh>
+#include <EnergyPlus/UtilityRoutines.hh>
```
Why is this being added here?

amirroth commented on src/EnergyPlus/DXCoils.hh:
```diff
@@ -345,8 +347,7 @@ namespace DXCoils {
         // for multispeed DX coil type
-        std::string FuelType;                       // Fuel type string
-        Constant::ResourceType FuelTypeNum;         // Fuel type number
+        Constant::eResource FuelTypeNum;            // Fuel type number
```
Rename this into `FuelType` (removing `Num`).

amirroth commented on src/EnergyPlus/DataGlobalConstants.cc:
👏

amirroth commented on src/EnergyPlus/DataGlobalConstants.hh:
```diff
@@ -114,21 +126,257 @@ namespace Constant {
+        eResourceNames[static_cast<int>(ePollutant2eResource[static_cast<int>(ePollutant::Pb)])],
+        eResourceNames[static_cast<int>(ePollutant2eResource[static_cast<int>(ePollutant::NuclearHigh)])],
+        eResourceNames[static_cast<int>(ePollutant2eResource[static_cast<int>(ePollutant::NuclearLow)])]};
+
```
Very nice. All of these functions converted to `constexpr` array lookups.

amirroth commented on src/EnergyPlus/HeatingCoils.cc:
```diff
@@ -836,7 +837,7 @@ namespace HeatingCoils {
             auto &heatingCoilNumericFields = state.dataHeatingCoils->HeatingCoilNumericFields(CoilNum);
             CurrentModuleObject = "Coil:Heating:Gas:MultiStage";
-            heatingCoil.FuelType_Num = Constant::ResourceType::Natural_Gas;
+            heatingCoil.ResourceType = Constant::eResource::NaturalGas;
```
Shouldn't this be a `FuelType` and not `ResourceType`? What non-fuel resources are valid here?

Myoldmopar replied:
@dareumnam I feel like this PR is ready to, and needs to, get merged in to avoid conflicts with other refactors.  If this is still a point to be addressed, please add it to the next round.

amirroth commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -1437,166 +1437,39 @@ namespace OutputProcessor {
-            } else if (meterType == "OTHERFUEL2") { // other fuel type (defined by user)
-                OutResourceType = "OtherFuel2";
-
+        std::string const &meterType = UtilityRoutines::MakeUPPERCase(UserInputResourceType);
```
`MakeUPPERCase` returns a string object so this should not be a `&`.

amirroth commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -1437,166 +1437,39 @@ namespace OutputProcessor {
+        int eMeterResource = getEnumerationValue(Constant::eResourceNamesUC, meterType);
+
+        if (eMeterResource == static_cast<int>(Constant::eResource::Invalid)) {
+            if (meterType == "H2O") {
```
Is it critical that we continue to allow these variants?

mjwitte replied:
> Is it critical that we continue to allow these variants?

If this is always interpreting an input \choice field, then there's no need to support variants, because they won't get past the input processor.

amirroth replied:
Halelujah!

dareumnam replied:
This function interprets three things:
1. input \choice field of `Meter:Custom `
2. input \choice field of `Meter:CustomDecrement`
3. `ResourceTypeKey` argument of `SetupOutputVariable()`. 

1 and 2 are input \choiced fields. As for 3,  I have confirmed that the `ResourceTypeKey`s of `SetupOutputVariable()`s in each object are properly preset and these variants are not being used. So it seems like we don't need to continue to allow these variants.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -8391,7 +8391,7 @@ void DefineIdealCondEntSetPointManager::SetupMeteredVarsForSetPt(EnergyPlusData
     unitsForVar.allocate(NumVariables);
 
     for (int varN = 1; varN <= NumVariables; ++varN) {
-        ResourceTypes.insert(std::pair<int, Constant::ResourceType>(varN, Constant::ResourceType::None));
```
This is a bad use of `std::map` because the key (`varN`) is a dense integer set.

Myoldmopar replied:
@dareumnam another item to do in the next round, or if not, then explain why it's not being changed.

dareumnam replied:
Oops, I missed this one. I'll address this on the follow-up PR.

amirroth reviewed (commented):
Looks quite good overall. A few immediate simplifications are possible, primarily removing `MakeUPPERCase` and checks for `X::Invalid` for required fields. One open question is whether we should change `FuelType` fields from `eResource` to the smaller `eFuel` in this PR or in a follow-up PR.

Myoldmopar commented:
This has got a chunk of conflicts to resolve after the merge binge of today.  I suspect they are all small and straightforward, but there are several nonetheless.  I think there are also a few comments to address.  I'll keep this one aside for now, but I'm happy to help work on resolving conflicts later.

dareumnam committed:
pull develop and merge conflicts

dareumnam committed:
remove MakeUPPERCase in variables from input processor & remove error messages for required field

dareumnam committed:
clang format

dareumnam committed:
remove Num

dareumnam committed:
remove variants of resource type name in GetStandardMeterResourceType()

dareumnam committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into FuelTypeNames

dareumnam commented:
All comments are addressed, this PR is ready to get merged in.  I'll keep converting all FuelType variables from `eResource` to `eFuel` on a new branch.

Myoldmopar commented:
This is great @dareumnam, thanks!  All good on CI, so I'm just going to do a quick local build while reviewing the final set of changes.

Myoldmopar commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -195,17 +196,8 @@ namespace BoilerSteam {
-                ErrorsFound = true;
-            }
+            thisBoiler.FuelType =
+                static_cast<Constant::eResource>(getEnumerationValue(Constant::eResourceNamesUC, state.dataIPShortCut->cAlphaArgs(2)));
```
:+1:

Myoldmopar commented on src/EnergyPlus/EconomicLifeCycleCost.hh:
```diff
@@ -356,41 +356,41 @@ namespace EconomicLifeCycleCost {
 
         // Default Constructor
-        UsePriceEscalationType() : resource(Constant::ResourceType::None), escalationStartYear(0), escalationStartMonth(0)
+        UsePriceEscalationType() : resource(Constant::eResource::Invalid), escalationStartYear(0), escalationStartMonth(0)
```
@dareumnam on the next PR round, could you just remove these default constructors, and instead simply put the assignments up where the variables are declared a few lines up?

dareumnam replied:
Will do!

Myoldmopar commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -1437,166 +1437,25 @@ namespace OutputProcessor {
-                ShowSevereError(state,
-                                format("GetStandardMeterResourceType: Illegal OutResourceType (for Meters) Entered={}", UserInputResourceType));
-                ErrorsFound = true;
-            }
```
Yay!

Myoldmopar commented on tst/EnergyPlus/unit/CMakeLists.txt:
```diff
@@ -79,7 +79,6 @@ set(test_src
     DOASEffectOnZoneSizing.unit.cc
     DXCoils.unit.cc
     DataEnvironment.unit.cc
-    DataGlobalConstants.unit.cc
```
Yay!

Myoldmopar reviewed (approved):
I don't think there is anything left here that should hold this up.  We can certainly discuss variable naming and fuel/resource concerns, but in general this is a really great, and relatively large, set of changes and should merge if it's all happy.  Thanks @dareumnam

Myoldmopar commented:
OK, everything is happy locally.  Just make sure straggling comments are addressed or discussed on the follow-up PR, and we are good to go.  Merging.  Thanks @dareumnam plus all reviewers, this has been a long time coming.
